src/main/java/com/google/firebase/messaging/Message.java

```diff
@@ -0,0 +1,225 @@
+ * limitations under the License.
+ */
+
+package com.google.firebase.messaging;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.api.client.util.Key;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.primitives.Booleans;
+import com.google.firebase.internal.NonNull;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Represents a message that can be sent via Firebase Cloud Messaging (FCM). Contains payload
+ * information as well as the recipient information. In particular, the message must contain
+ * exactly one of token, topic or condition parameters. Instances of this class are thread-safe
+ * and immutable. Use {@link Message.Builder} co crete new instances.
+ *
+ * @see
+ * <a href="https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages">FCM message
+ * format</a>
+ */
+public class Message {
+
+  @Key("data")
+  private final Map<String, String> data;
+
+  @Key("notification")
+  private final Notification notification;
+
+  @Key("android")
+  private final AndroidConfig androidConfig;
+
+  @Key("webpush")
+  private final WebpushConfig webpushConfig;
+
+  @Key("apns")
+  private final ApnsConfig apnsConfig;
+
+  @Key("token")
+  private final String token;
+
+  @Key("topic")
+  private final String topic;
+
+  @Key("condition")
+  private final String condition;
+
+  private Message(Builder builder) {
+    this.data = builder.data.isEmpty() ? null : ImmutableMap.copyOf(builder.data);
+    this.notification = builder.notification;
+    this.androidConfig = builder.androidConfig;
+    this.webpushConfig = builder.webpushConfig;
+    this.apnsConfig = builder.apnsConfig;
+    int count = Booleans.countTrue(
+        !Strings.isNullOrEmpty(builder.token),
+        !Strings.isNullOrEmpty(builder.topic),
+        !Strings.isNullOrEmpty(builder.condition)
+    );
+    checkArgument(count == 1, "Exactly one of token, topic or condition must be specified");
+    if (builder.topic != null) {
+      checkArgument(!builder.topic.startsWith("/topics/"),
+          "Topic name must not contain the /topics/ prefix");
+      checkArgument(builder.topic.matches("[a-zA-Z0-9-_.~%]+"), "Malformed topic name");
+    }
+    this.token = builder.token;
+    this.topic = builder.topic;
+    this.condition = builder.condition;
+  }
+
+  /**
+   * Creates a new {@link Message.Builder}.
+   *
+   * @return A {@link Message.Builder} instance.
+   */
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class  Builder {
+
+    private final Map<String, String> data = new HashMap<>();
+    private Notification notification;
+    private AndroidConfig androidConfig;
+    private WebpushConfig webpushConfig;
+    private ApnsConfig apnsConfig;
+    private String token;
+    private String topic;
+    private String condition;
+
+    private Builder() {
+
```

Comment:
remove empty line, or even just
`private Builder() {}`
here and elsewhere

Reply:
Done